File: modules/daf/daf.lua
-- SPDX-License-Identifier: GPL-3.0-or-later

-- Load dependent modules
if not view then modules.load('view') end
if not policy then modules.load('policy') end

-- Actions
local actions = {
	pass = function() return policy.PASS end,
	deny = function () return policy.DENY end,
	drop = function() return policy.DROP end,
	tc = function() return policy.TC end,
	truncate = function() return policy.TC end,
	forward = function (g)
		local addrs = {}
		local tok = g()
		for addr in string.gmatch(tok, '[^,]+') do
			table.insert(addrs, addr)
		end
		return policy.FORWARD(addrs)
	end,
	mirror = function (g)
		return policy.MIRROR(g())
	end,
	reroute = function (g)
		local rules = {}
		local tok = g()
		while tok do
			local from, to = tok:match '([^-]+)-(%S+)'
			rules[from] = to
			tok = g()
		end
		return policy.REROUTE(rules)
	end,
	rewrite = function (g)
		local rules = {}
		local tok = g()
		while tok do
			-- This is currently limited to A/AAAA rewriting
			-- in fixed format '<owner> <type> <addr>'
			local _, to = g(), g()
			rules[tok] = to
			tok = g()
		end
		return policy.REROUTE(rules, true)
	end,
}

-- Filter rules per column
local filters = {
	-- Filter on QNAME (either pattern or suffix match)
	qname = function (g)
		local op, val = g(), todname(g())
		if     op == '~' then return policy.pattern(true, val:sub(2)) -- Skip leading label length
		elseif op == '=' then return policy.suffix(true, {val})
		else error(string.format('invalid operator "%s" on qname', op)) end
	end,
	-- Filter on source address
	src = function (g)
		local op = g()
		if op ~= '=' then error('address supports only "=" operator') end
		return view.rule_src(true, g())
	end,
	-- Filter on destination address
	dst = function (g)
		local op = g()
		if op ~= '=' then error('address supports only "=" operator') end
		return view.rule_dst(true, g())
	end,
}

local function parse_filter(tok, g, prev)
	if not tok then error(string.format('expected filter after "%s"', prev)) end
	local filter = filters[tok:lower()]
	if not filter then error(string.format('invalid filter "%s"', tok)) end
	return filter(g)
end

local function parse_rule(g)
	-- Allow action without filter
	local tok = g()
	if tok == nil then
		error('empty rule is not allowed')
	end
	if not filters[tok:lower()] then
		return tok, nil
	end
	local f = parse_filter(tok, g)
	-- Compose filter functions on conjunctions
	-- or terminate filter chain and return
	tok = g()
	while tok do
		if tok:lower() == 'and' then
			local fa, fb = f, parse_filter(g(), g, tok)
			f = function (req, qry) return fa(req, qry) and fb(req, qry) end
		elseif tok:lower() == 'or' then
			local fa, fb = f, parse_filter(g(), g, tok)
			f = function (req, qry) return fa(req, qry) or fb(req, qry) end
		else
			break
		end
		tok = g()
	end
	return tok, f
end

local function parse_query(g)
	local ok, actid, filter = pcall(parse_rule, g)
	if not ok then return nil, actid end
	actid = actid:lower()
	if not actions[actid] then return nil, string.format('invalid action "%s"', actid) end
	-- Parse and interpret action
	local action = actions[actid]
	if type(action) == 'function' then
		action = action(g)
	end
	return actid, action, filter
end

-- Compile a rule described by query language
-- The query language is modelled by iptables/nftables
-- conj = AND | OR
-- op = IS | NOT | LIKE | IN
-- filter = <key> <op> <expr>
-- rule = <filter> | <filter> <conj> <rule>
-- action = PASS | DENY | DROP | TC | FORWARD
-- query = <rule> <action>
local function compile(query)
	local g = string.gmatch(query, '%S+')
	return parse_query(g)
end

-- @function Describe given rule for presentation
local function rule_info(r)
	return {info=r.info, id=r.rule.id, active=(r.rule.suspended ~= true), count=r.rule.count}
end

-- Module declaration
local M = {
	rules = {}
}

-- @function Remove a rule

-- @function Cleanup module
function M.deinit()
	if http then
		local endpoints = http.configs._builtin.webmgmt.endpoints
		endpoints['/daf'] = nil
		endpoints['/daf.js'] = nil
		http.snippets['/daf'] = nil
	end
end

-- @function Add rule
function M.add(rule)
	-- Ignore duplicates
	for _, r in ipairs(M.rules) do
		if r.info == rule then return r end
	end
	local id, action, filter = compile(rule)
	if not id then error(action) end
	-- Combine filter and action into policy
	local p
	if filter then
		p = function (req, qry)
			return filter(req, qry) and action
		end
	else
		p = function ()
			return action
		end
	end
	local desc = {info=rule, policy=p}
	-- Enforce in policy module, special actions are postrules
	if id == 'reroute' or id == 'rewrite' then
		desc.rule = policy.add(p, true)
	else
		desc.rule = policy.add(p)
	end
	table.insert(M.rules, desc)
	return desc
end

-- @function Remove a rule
function M.del(id)
	for key, r in ipairs(M.rules) do
		if r.rule.id == id then
			policy.del(id)
			table.remove(M.rules, key)
			return true
		end
	end
	return nil
end

-- @function Find a rule
function M.get(id)
	for _, r in ipairs(M.rules) do
		if r.rule.id == id then
			return r
		end
	end
	return nil
end

-- @function Enable/disable a rule
function M.toggle(id, val)
	for _, r in ipairs(M.rules) do
		if r.rule.id == id then
			r.rule.suspended = not val
			return true
		end
	end
	return nil
end

-- @function Enable/disable a rule
function M.disable(id)
	return M.toggle(id, false)
end
function M.enable(id)
	return M.toggle(id, true)
end

local function consensus(op, ...)
	local ret = false
	local results = map(string.format(op, ...))
	for idx, r in ipairs(results) do
		if idx == 1 then
			-- non-empty table, init to true
			ret = true
		end
		ret = ret and r
	end
	return ret
end

-- @function Public-facing API
local function api(h, stream)
	local m = h:get(':method')
	-- GET method
	if m == 'GET' then
		local path = h:get(':path')
		local id = tonumber(path:match '/([^/]*)$')
		if id then
			local r = M.get(id)
			if r then
				return rule_info(r)
			end
			return 404, '"No such rule"' -- Not found
		else
			local ret = {}
			for _, r in ipairs(M.rules) do
				table.insert(ret, rule_info(r))
			end
			return ret
		end
	-- DELETE method
	elseif m == 'DELETE' then
		local path = h:get(':path')
		local id = tonumber(path:match '/([^/]*)$')
		if id then
			if consensus('daf.del(%s)', id) then
				return tojson(true)
			end
			return 404, '"No such rule"' -- Not found
		end
		return 400 -- Request doesn't have numeric id
	-- POST method
	elseif m == 'POST' then
		local query = stream:get_body_as_string()
		if query then
			local ok, r = pcall(M.add, query)
			if not ok then return 500, string.format('"%s"', r:match('/([^/]+)$')) end
			-- Dispatch to all other workers:
			-- we ignore return values except error() because they are not serializable
			consensus('daf.add "%s" and true', query)
			return rule_info(r)
		end
		return 400
	-- PATCH method
	elseif m == 'PATCH' then
		local path = h:get(':path')
		local id, action, val = path:match '(%d+)/([^/]*)/([^/]*)$'
		id = tonumber(id)
		if not id or not action or not val then
			return 400 -- Request not well formatted
		end
		-- We do not support more actions
		if action == 'active' then
			if consensus('daf.toggle(%d, %s)', id, val == 'true' or 'false') then
				return tojson(true)
			else
				return 404, '"No such rule"'
			end
		else
			return 501, '"Action not implemented"'
		end
	end
end

local function getmatches()
	local update = {}
	-- Must have string keys for JSON object and not an array
	local inst_counters = map('ret = {} '
		.. 'for _, rule in ipairs(daf.rules) do '
			.. 'ret[tostring(rule.rule.id)] = rule.rule.count '
		.. 'end '
		.. 'return ret')
	for inst_idx=1, inst_counters.n do
		for r_id, r_cnt in pairs(inst_counters[inst_idx]) do
			update[r_id] = (update[r_id] or 0) + r_cnt
		end
	end
	return update
end

-- @function Publish DAF statistics
local function publish(_, ws)
	local ok, last = true, nil
	while ok do
		-- Check if we have new rule matches
		local diff = {}
		local has_update, update = pcall(getmatches)
		if has_update then
			if last then
				for id, count in pairs(update) do
					if not last[id] or last[id] < count then
						diff[id] = count
					end
				end
			end
			last = update
		end
		-- Update counters when there is a new data
		if next(diff) ~= nil then
			ok = ws:send(tojson(diff))
		else
			ok = ws:send_ping()
		end
		worker.sleep(1)
	end
end

function M.init()
	-- avoid ordering problem between HTTP and daf module
	event.after(0, M.config)
end

-- @function Configure module
function M.config()
	if not http then
		if verbose() then
			log('[daf ] HTTP API unavailable because HTTP module is not loaded, use modules.load("http")')
		end
		return
	end
	local endpoints = http.configs._builtin.webmgmt.endpoints
	-- Export API and data publisher
	endpoints['/daf.js'] = http.page('daf.js', 'daf')
	endpoints['/daf'] = {'application/json', api, publish}
	-- Export snippet
	http.snippets['/daf'] = {'Application Firewall', [[
		<script type="text/javascript" src="daf.js"></script>
		<div class="row" style="margin-bottom: 5px">
			<form id="daf-builder-form">
				<div class="col-md-11">
					<input type="text" id="daf-builder" class="form-control" aria-label="..." />
				</div>
				<div class="col-md-1">
					<button type="button" id="daf-add" class="btn btn-default btn-sm">Add</button>
				</div>
			</form>
		</div>
		<div class="row">
			<div class="col-md-12">
				<table id="daf-rules" class="table table-striped table-responsive">
				<th><td>No rules here yet.</td></th>
				</table>
			</div>
		</div>
	]]}
end

return M

File: modules/daf/daf.js
/* Filter grammar
 * SPDX-License-Identifier: GPL-3.0-or-later */
const dafg = {
	key: {'qname': true, 'src': true, 'dst': true},
	op: {'=': true, '~': true},
	conj: {'and': true, 'or': true},
	action: {'pass': true, 'deny': true, 'drop': true, 'truncate': true, 'forward': true, 'reroute': true, 'rewrite': true, 'mirror': true},
	suggest: [
		'QNAME = example.com',
		'QNAME ~ %d+.example.com',
		'SRC = 127.0.0.1',
		'SRC = 127.0.0.1/8',
		'DST = 127.0.0.1',
		'DST = 127.0.0.1/8',
		/* Action examples */
		'PASS', 'DENY', 'DROP', 'TRUNCATE',
		'FORWARD 127.0.0.1',
		'MIRROR 127.0.0.1',
		'REROUTE 127.0.0.1-192.168.1.1',
		'REROUTE 127.0.0.1/24-192.168.1.0',
		'REWRITE example.com A 127.0.0.1',
		'REWRITE example.com AAAA ::1',
	]
};

function setValidateHint(cls) {
	var builderForm = $('#daf-builder-form');
	builderForm.removeClass('has-error has-warning has-success');
	if (cls) {
		builderForm.addClass(cls);
	}
}

function validateToken(tok, tbl) {
	if (tok.length > 0 && tok[0].length > 0) {
		if (tbl[tok[0].toLowerCase()]) {
			setValidateHint('has-success');
			return true;
		} else { setValidateHint('has-error'); }
	} else { setValidateHint('has-warning'); }
	return false;
}

function parseOption(tok) {
	var key = tok.shift().toLowerCase();
	var op = null;
	if (dafg.key[key]) {
		op = tok.shift();
		if (op) {
			op = op.toLowerCase();
		}
	}
	const item = {
		text: key.toUpperCase() + ' ' + (op ? op.toUpperCase() : '') + ' ' + tok.join(' '),
	};
	if (dafg.key[key]) {
		item.class = 'tag-default';
	} else if (dafg.action[key]) {
		item.class = 'tag-warning';
	} else if (dafg.conj[key]) {
		item.class = 'tag-success';
	}
	return item;
}

function createOption(input) {
	const item = parseOption(input.split(' '));
	item.value = input;
	return item;
}

function dafComplete(form) {
	const items = form.items;
	for (var i in items) {
		const tok = items[i].split(' ')[0].toLowerCase();
		if (dafg.action[tok]) {
			return true;
		}
	}
	return false;
}

function formatRule(input) {
	const tok = input.split(' ');
	var res = [];
	while (tok.length > 0) {
		const key = tok.shift().toLowerCase();
		if (dafg.key[key]) {
			var item = parseOption([key, tok.shift(), tok.shift()]);
			res.push('<span class="label tag '+item.class+'">'+item.text+'</span>');
		} else if (dafg.action[key]) {
			var item = parseOption([key].concat(tok));
			res.push('<span class="label tag '+item.class+'">'+item.text+'</span>');
			tok.splice(0, tok.length);
		} else if (dafg.conj[key]) {
			var item = parseOption([key]);
			res.push('<span class="label tag '+item.class+'">'+item.text+'</span>');
		}
	}
	return res.join('');
}

function toggleRule(row, span, enabled) {
	if (!enabled) {
		span.removeClass('glyphicon-pause');
		span.addClass('glyphicon-play');
		row.addClass('warning');
	} else {
		span.removeClass('glyphicon-play');
		span.addClass('glyphicon-pause');
		row.removeClass('warning');
	}
}

function ruleControl(cell, type, url, action) {
	const row = cell.parent();
	$.ajax({
		url: 'daf/' + row.data('rule-id') + url,
		type: type,
		success: action,
		fail: function (data) {
			row.show();
			const reason = data.responseText.length > 0 ? data.responseText : 'internal error';
			cell.find('.alert').remove();
			cell.append(
				'<div class="alert alert-danger" role="alert">'+
				'Failed (code: '+data.status+', reason: '+reason+').'+
				'</div>'
			);
		},
	});
}

function bindRuleControl(cell) {
	const row = cell.parent();
	cell.find('.daf-remove').click(function() {
		row.hide();
		ruleControl(cell, 'DELETE', '', function (data) {
			cell.parent().remove();
		});
	});
	cell.find('.daf-suspend').click(function() {
		const span = $(this).find('span');
		ruleControl(cell, 'PATCH', span.hasClass('glyphicon-pause') ? '/active/false' : '/active/true');
		toggleRule(row, span, span.hasClass('glyphicon-play'));
	});
}

function loadRule(rule, tbl) {
	const row = $('<tr data-rule-id="'+rule.id+'" />');
	row.append('<td class="daf-rule">' + formatRule(rule.info) + '</td>');
	row.append('<td class="daf-count">' + rule.count + '</td>');
	row.append('<td class="daf-rate"><span class="badge"></span></td>');
	row.append('<td class="daf-ctl text-right">' +
		'<div class="btn-group btn-group-xs">' +
		'<button class="btn btn-default daf-suspend"><span class="glyphicon" aria="hidden" /></button>' +
		'<button class="btn btn-default daf-remove"><span class="glyphicon glyphicon-remove" aria="hidden" /></button>' +
		'</div></td>');
	tbl.append(row);
	/* Bind rule controls */
	bindRuleControl(row.find('.daf-ctl'));
	toggleRule(row, row.find('.daf-suspend span'), rule.active);
}

/* Load the filter table from JSON */
function loadTable(resp) {
	const tbl = $('#daf-rules')
	tbl.children().remove();
	tbl.append('<tr><th>Rule</th><th>Matches</th><th>Rate</th><th></th></tr>')
	for (var i in resp) {
		loadRule(resp[i], tbl);
	}
}

$(function() {
	/* Load the filter table. */
	$.ajax({
		url: 'daf',
		type: 'get',
		dataType: 'json',
		success: loadTable
	});
	/* Listen for counter updates */
	const wsStats = (secure ? 'wss://' : 'ws://') + location.host + '/daf';
	const ws = new Socket(wsStats);
	var lastRateUpdate = Date.now();
	ws.onmessage = function(evt) {
		var data = JSON.parse(evt.data);
		/* Update heartbeat clock */
		var now = Date.now();
		var dt = now - lastRateUpdate;
		lastRateUpdate = now;
		/* Update match counts and rates */
		$('#daf-rules .daf-rate span').text('');
		for (var key in data) {
			const row = $('tr[data-rule-id="'+key+'"]');
			if (row) {
				const cell = row.find('.daf-count');
				const diff = data[key] - parseInt(cell.text());
				cell.text(data[key]);
				const badge = row.find('.daf-rate span');
				if (diff > 0) {
					/* Normalize difference to heartbeat (in msecs) */
					const rate = Math.ceil((1000 * diff) / dt);
					badge.text(rate + ' pps');
				}
			}
		}
	};
	/* Rule builder UI */
	$('#daf-builder').selectize({
		delimiter: ',',
		persist: true,
		highlight: true,
		closeAfterSelect: true,
		onItemAdd: function (input, item) {
		    setValidateHint();
		    /* Prevent new rules when action is specified */
		    const tok = input.split(' ');
		    if (dafg.action[tok[0].toLowerCase()]) {
		    	$('#daf-add').focus();
		    } else if(dafComplete(this)) {
		    	/* No more rules after query is complete. */
		    	item.remove();
		    }
		},
		createFilter: function (input) {
			const tok = input.split(' ');
			var key, op, expr;
			/* If there are already filters, allow conjunctions. */
			if (tok.length > 0 && this.items.length > 0 && dafg.conj[tok[0]]) {
				setValidateHint();
				return true;
			}
			/* First token is expected to be filter key,
			 * or any postrule with a parameter */
			if (validateToken(tok, dafg.key)) {
				key = tok.shift();
			} else if (tok.length > 1 && validateToken(tok, dafg.action)) {
				setValidateHint();
				return true;
			} else {
				return false;
			}
			/* Input is a filter - second token must be operator */
			if (validateToken(tok, dafg.op)) {
				op = tok.shift();
			} else {
				return false;
			}
			/* Input is a filter - the rest of the tokens are RHS arguments. */
			if (tok.length > 0 && tok[0].length > 0) {
				expr = tok.join(' ');
			} else {
				setValidateHint('has-warning');
				return false;
			}
			setValidateHint('has-success');
			return true;
		},
		create: createOption,
		render: {
			item: function(item, escape) {
				return '<div class="name '+item.class+'">' + escape(item.text) + '</span>';
			},
		},
	});
	/* Add default suggestions. */
	const dafBuilder = $('#daf-builder')[0].selectize;
	for (var i in dafg.suggest) {
		dafBuilder.addOption(createOption(dafg.suggest[i]));
	}
	/* Rule builder submit */
	$('#daf-add').click(function () {
		const form = $('#daf-builder-form').parent();
		if (dafBuilder.items.length == 0 || form.hasClass('has-error')) {
			return;
		}
		/* Clear previous errors and resubmit. */
		form.parent().find('.alert').remove();
		$.post('daf', dafBuilder.items.join(' '))
			.done(function (data) {
				dafBuilder.clear();
				loadRule(data, $('#daf-rules'));
			})
			.fail(function (data) {
				const reason = data.responseText.length > 0 ? data.responseText : 'internal error';
				form.after(
					'<div class="alert alert-danger" role="alert">'+
				       'Couldn\'t add rule (code: '+data.status+', reason: '+reason+').'+
				    '</div>'
				);
			});
	});
});
